In [1]:
import pandas as pd

# Import dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Check for missing values
missing_values = df.isna().sum()

# Print missing values count
print(missing_values)


row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


In [2]:
import pandas as pd

# Load the dataset from the provided URL
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
villagers_data = pd.read_csv(url)

# Print the number of rows and columns
print(villagers_data.shape)


(391, 11)


2.2 'observation' is rows in the dataset, in the dataset i used means each villager. 'variables' is more specific than the observation it is
it present each specific features in the observation

In [3]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
villagers_data = pd.read_csv(url)

# Get a summary of numeric columns
print(villagers_data.describe())

# Count unique values in a categorical column (e.g., species)
print(villagers_data['species'].value_counts())


            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64


In [3]:
import pandas as pd

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Get the shape of the dataset
print("Shape of the dataset:", titanic_data.shape)

# Summary of numeric columns
print("\nSummary of numeric columns:")
print(titanic_data.describe())

# Check for non-numeric columns
print("\nNon-numeric columns:")
print(titanic_data.select_dtypes(include='object').columns)

# Check for missing values in numeric columns
print("\nMissing values in numeric columns:")
print(titanic_data.isnull().sum())


Shape of the dataset: (891, 15)

Summary of numeric columns:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Non-numeric columns:
Index(['sex', 'embarked', 'class', 'who', 'deck', 'embark_town', 'alive'], dtype='object')

Missing values in numeric columns:
survived         0
pclass           0
sex              0
age            177
sibsp       

"attribute" give you back information directly from the row dataset, it don't need to compute, "method" do some calculations and give you back the result. for example, titanic_data.shape return the row and columns of the dataset, and titanic_data.describe() return the summary data of the data sat.

6.
Count:

Definition: The number of non-missing values in the column. It tells you how many valid entries are present in that column.
Mean:

Definition: The average value of the data in the column. It's calculated by summing all the values and then dividing by the count of non-missing values.
Std (Standard Deviation):

Definition: A measure of the amount of variation or dispersion in the column’s values. A low standard deviation means that values tend to be close to the mean, while a high standard deviation indicates that values are spread out over a wider range.
Min:

Definition: The smallest value in the column.
25% (25th Percentile or First Quartile):

Definition: The value below which 25% of the data falls. It divides the lowest 25% of the data from the remaining 75%.
50% (50th Percentile or Median):

Definition: The middle value of the data. Half of the data falls below this value, and half is above it.
75% (75th Percentile or Third Quartile):

Definition: The value below which 75% of the data falls. It divides the lowest 75% of the data from the remaining 25%.
Max:

Definition: The largest value in the column.

7-1
if you want to remove rows that contain missing values, df.dropna() might be peferred over using del df['col']
for example:

In [2]:
import pandas as pd

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
titanic_data = pd.read_csv(url)

# Step 1: Check missing values before cleaning
print("Missing values before using df.dropna():")
print(titanic_data.isnull().sum())

# Step 2: Drop rows with missing values
cleaned_data = titanic_data.dropna()

# Step 3: Check missing values after using df.dropna()
print("\nMissing values after using df.dropna():")
print(cleaned_data.isnull().sum())

# Step 4: Check the shape before and after cleaning
print("\nShape before cleaning:", titanic_data.shape)
print("Shape after cleaning:", cleaned_data.shape)


Missing values before using df.dropna():
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Missing values after using df.dropna():
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

Shape before cleaning: (891, 15)
Shape after cleaning: (182, 15)


7-2
If there is a column with a lot of data missed, you can use del df['col'] might be preferred over using df.dropna(), for example, in the same data site there is a column "cabin" it has a lot of missing values, and don't have much analysis value, its better to remove the whole column instead of removing the rows which might cause loss of data. 
for example:

In [8]:
import pandas as pd

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Show the first few rows to see the columns and missing values
print(df.head())

# Check how many missing values there are in each column
print(df.isna().sum())

# Use del to completely remove the 'deck' column since it has many missing values and might not be relevant
del df['deck']

# Print the DataFrame columns after deletion
print(df.columns)


   survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck      

7-3
this can prevent necessary row got deleted by df.dropna() because if there are only 1 missing value in that row, df.dropna() will delete the whole row, by doing del df['col'] can delete the unnecessary column where it has the most missed data.

In [10]:
import pandas as pd

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Display initial row and column count
print("Initial shape:", df.shape)

# Check how many missing values there are in each column
print(df.isna().sum())

# Delete 'deck' column before dropping rows with missing values
del df['deck']

# Now, drop rows with any missing values
df_cleaned = df.dropna()

# Display row and column count after cleaning
print("Shape after cleaning:", df_cleaned.shape)



Initial shape: (891, 15)
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64
Shape after cleaning: (712, 14)


7-4
approach:
1.check which column has the most missing values, if it have too many missing value it might not be useful, so delete it using del df['']
2.clean remaining missing data, use df.dropna() clean the rows with missing values.

before:
Shape: (891 rows, 15 columns)
Missing values per column:
deck: 688 missing values
age: 177 missing values
embarked: 2 missing values
embark_town: 2 missing value
after:
Final shape: (712 rows, 14 columns)
Missing values per column: No missing values, as all missing data has been removed.

Justification for Approach:
i choose to use del df['col'] before df.dropna() because the column "deck" have 688 missing value, if i choose to use df.dropna() before deleting the"deck" column will cause 77% of rows removed. 

In [11]:
import pandas as pd

# Load the Titanic dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Initial shape and missing values
print("Initial shape:", df.shape)
print("Missing values before cleaning:")
print(df.isna().sum())

# Step 1: Remove the 'deck' column since it has too many missing values (more than 75%)
del df['deck']

# Step 2: Drop rows with missing values in remaining important columns
df_cleaned = df.dropna()

# Final shape and missing values
print("\nFinal shape after cleaning:", df_cleaned.shape)
print("Missing values after cleaning:")
print(df_cleaned.isna().sum())


Initial shape: (891, 15)
Missing values before cleaning:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

Final shape after cleaning: (712, 14)
Missing values after cleaning:
survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64


8-1
df.groupby("col1"), this groups every unique value in one group and form a seprate column,
["col2"], this select the choosen column and form a table with col1, for example, ("income") shows the income of the person in a dataset, ["age"] is the column of person's age in the data set, and the table will be about the stat of people's income when they at specific age.
.describe() will show the summary of stats in a table about col2 and col1, if the col2 is numerical it out puts details: count, mean, std, min, 25%, 50%, 75%, max. if the col2 is non numerical it will show the count, unique top and freq.


In [8]:
import pandas as pd

column_names = [
    'age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status',
    'occupation', 'relationship', 'race', 'sex', 'capital_gain', 'capital_loss',
    'hours_per_week', 'native_country', 'income'
]

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data'
df = pd.read_csv(url, header=None, names=column_names, na_values=' ?')

grouped_description = df.groupby("income")["age"].describe()
print(grouped_description)

NameError: name 'age' is not defined

8-2
It's fundamentally different because in "df.describe" count show the non missing values in each columns, treating columns independently. In
df.groupby("col1")["col2"].describe(), it shows how many values in col2 exist for each unique group in col1.

8-3
errors:
A) NameError: name 'pd' is not defined
B)FileNotFoundError: [Errno 2] No such file or directory: 'titanic.csv'
C)NameError: name 'DF' is not defined
D)SyntaxError: '(' was never closed
E)AttributeError: 'DataFrame' object has no attribute 'group_by' OR AttributeError: 'DataFrame' object has no attribute 'describle'
F)KeyError: 'Sex'or KeyError: 'Age'
G)NameError: name 'income' is not defined OR NameError: name 'age' is not defined
I don't think google search is a faster way than chatgpt in terms of fixing errors, by typing the error to chatbot, it can directly tell you where is wrong with the code, but in google search you need to find the corresponding one among a lot of site

NameError: name 'pd' is not defined

summary:
df = pd.read_csv(url) Loads the dataset from a URL into a pandas DataFrame.
df.shape: Returns the dimensions (rows, columns) of the DataFrame.
df.describe(): Provides summary statistics (count, mean, std, etc.) for numerical columns.
df['column'].value_counts(): Returns the count of unique values in the specified column.
df.dropna(): Removes rows with any missing values (NaN) from the DataFrame.
del df['col']: Deletes the specified column from the DataFrame.
df.groupby("col1")["col2"].describe(): Groups the data by values in col1 and returns descriptive statistics for col2 within each group.

chatlog:
https://chatgpt.com/share/66e39a2b-8284-8003-a0b9-028c1027ffd1
https://chatgpt.com/share/66e39a41-e860-8003-baf2-40205095cb9d
https://chatgpt.com/share/66e39a4e-bcbc-8003-a57f-8347cdceaaa6
https://chatgpt.com/share/66e39a5c-0f20-8003-bec6-9732b6aaa220
https://chatgpt.com/share/66e39a66-76f8-8003-9146-8d6c36d44cf0
